vaadin-spring/src/main/java/com/vaadin/flow/spring/scopes/VaadinRouteScope.java

```diff
@@ -0,0 +1,380 @@
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.context.ApplicationContext;
+import org.springframework.web.context.support.WebApplicationContextUtils;
+
+import com.vaadin.flow.component.ComponentEventListener;
+import com.vaadin.flow.component.ComponentUtil;
+import com.vaadin.flow.component.DetachEvent;
+import com.vaadin.flow.component.HasElement;
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.component.page.ExtendedClientDetails;
+import com.vaadin.flow.router.AfterNavigationEvent;
+import com.vaadin.flow.router.AfterNavigationListener;
+import com.vaadin.flow.router.BeforeEnterEvent;
+import com.vaadin.flow.router.BeforeEnterListener;
+import com.vaadin.flow.router.RouterLayout;
+import com.vaadin.flow.server.UIInitEvent;
+import com.vaadin.flow.server.UIInitListener;
+import com.vaadin.flow.server.VaadinContext;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.VaadinServletContext;
+import com.vaadin.flow.server.VaadinSession;
+import com.vaadin.flow.shared.Registration;
+import com.vaadin.flow.spring.SpringVaadinSession;
+import com.vaadin.flow.spring.annotation.RouteScopeOwner;
+
+/**
+ * Implementation of Spring's
+ * {@link org.springframework.beans.factory.config.Scope} that binds the beans
+ * to a component in the navigation chain. Registered by default as the scope "
+ * {@value #VAADIN_ROUTE_SCOPE_NAME}".
+ *
+ * @see com.vaadin.flow.spring.annotation.VaadinSessionScope
+ * @author Vaadin Ltd
+ * @since
+ *
+ */
+public class VaadinRouteScope extends AbstractScope implements UIInitListener {
+
+    public static final String VAADIN_ROUTE_SCOPE_NAME = "vaadin-route";
+
+    private static class RouteStoreWrapper implements Serializable {
+
+        private final VaadinSession session;
+
+        private final Registration sessionDestroyListenerRegistration;
+
+        private final Map<String, BeanStore> routeStores;
+
+        private RouteStoreWrapper(VaadinSession session) {
+            assert session.hasLock();
+            this.session = session;
+            routeStores = new HashMap<>();
+            if (session instanceof SpringVaadinSession) {
+                sessionDestroyListenerRegistration = null;
+                ((SpringVaadinSession) session)
+                        .addDestroyListener(event -> destroy());
+            } else {
+                sessionDestroyListenerRegistration = session.getService()
+                        .addSessionDestroyListener(event -> destroy());
+            }
+        }
+
+        private BeanStore getBeanStore(UI ui) {
+            assert session.hasLock();
+            ExtendedClientDetails details = ui.getInternals()
+                    .getExtendedClientDetails();
+            String key = getUIStoreKey(ui);
+            if (details == null) {
+                ui.getPage().retrieveExtendedClientDetails(
+                        det -> relocateStore(ui, key));
```

Comment:
If the UI is detached before this callback is run, will the UI be garbage collected ?

Reply:
`UI` is not stored anywhere.
Its ID or window name is used to store a data.

This is a valid comment for data stored per detached `UI` and this is fixed now.